biapol_utilities/label/_match_labels.py

```diff
@@ -1,88 +1,104 @@
+from ._matching_algorithms import max_similarity
+from ._filter_similarity_matrix import suppressed_similarity
 
-def match_labels_stack(label_stack, method=intersection_over_union_matrix, **kwargs):
-    """Match labels from subsequent slices with specified method
+
+def match_labels_stack(label_stack, **kwargs):
+    """
+    Match labels from subsequent slices with specified method
 
     Parameters
     ----------
     label_stack : 3D-array, int
         Stack of 2D label images to be stitched with axis order ZYX
-    method : str, optional
-        Method to be used for stitching the masks. The default is intersection_over_union_matrix with a stitch threshold of 0.25.
-        *stitch_threshold* : float
-            Threshold value for iou value above which two labels are considered identical.
-            The default value is 0.25
+    *metric*: callable, optional
+        Method to be used to generate a metric of similarity between labels
+        in subsequent slices. Must return a matrix with
+        `shape=(max(n, m), max(n, m))`, where n and m are the present labels
+        in two subsequent slices z and z+1. The values in the matrix must be
+        normalized to a range of [0, 1], where 0 and signify minimal or maximal
+        similarity between two labels.
+        Default is `intersection_over_union_matrix`
+    *filter*:ccallable, optional
+        Method to be used to filter values from the similarity matrix. This can
+        help to speed up the matching process if, fo instance, entries in the
+        similarity matrix below a defined threshold are set to zero. Default is
+        `suppressed_similarity(similiarity_matrix, threshold=0.3)`
+    *matching* : callable, optional
+        Method to be used for matching the labels. This function is supposed
+        to return a binary matrix with `shape=(n+1, m+1)` corresponding to
+        `match=1`, `no_match=0` of the `n` labels in a given slice and
+        `m` labels in the following slice. The default is `max_similarity`.
 
     Returns
     -------
     3D-array, int
-        Stack of stitched masks
+        Stack of stitched labels
     """
 
-    if method == intersection_over_union_matrix:
-        
-        # iterate over masks
-        for i in range(len(label_stack)-1):
-            label_stack[i+1] = match_labels(label_stack[i], label_stack[i+1],
-                                            method=method, **kwargs)
-            
+    # iterate over stack of label images
+    for i in range(len(label_stack)-1):
+        label_stack[i+1] = match_labels(label_stack[i], label_stack[i+1],
+                                        **kwargs)
+
     return label_stack
 
-def match_labels(label_image_x, label_image_y, method=intersection_over_union_matrix, **kwargs):
-    """Match labels in label_image_y with labels in label_image_x based on similarity
-    as defined by the passed method.
-    
+
+def match_labels(label_image_x, label_image_y, **kwargs):
```

Comment:
Same question as above. If these three parameters are used in this funciton, why are they not mention in the parameter list of the function?

Reply:
I thought about this a bit and changed it in the current scripts. One thing I don't know, though:
If a callable is called as an argument, e.g. `match_labels()` is called as follows:
```python
output = match_labels(image_x, image_y,
  method_metric=IoU,
  method_filter=suppressed_similarity,
  method_match=max_similarity)
``` 
How could I add a parameter to, for instance, `suppressed_similarity`, which takes the threshold parameter as input? Using `**kwargs` would allow to simply pass kwargs to any subfunction. Or would I just use kwargs nonetheless?

Reply:
Hey @jo-mueller ,

I think the most flexible solution is using helper functions. I wrote a short example here: https://github.com/BiAPoL/scripts/blob/main/passing_methods_with_parameters.ipynb

At the end, the goal is to have an API that is extensible by others (including ourselves), who do not have write-access to our code. That's why we need to do two things: 
a) keep our code simple
b) document it well, e.g. using examples

Let me know what you think!

Reply:
Ok, but these helper functions would best be simply introduced in an example notebook and note as actual functions in the library?  From how I understand the linked example, the threshold would be hardcoded in such a helper function?

Hence, maybe we should explain users how to define/write such helper functions rather than providing them ourselves?

Reply:
Hey Johannes @jo-mueller , I found a better way! Check out [partial](https://docs.python.org/3/library/functools.html#functools.partial). Shall I write a small notebook demonstrating it?

Reply:
Veeeeery cool method, I added it to the demonstration notebook.